### PR TITLE
HTML API: Fix tag processor token length

### PR DESCRIPTION
**Pull request by sirreal**
It seems like the open tags have incorrect length in the tag processor. I'd expect that `substr( $html, $token_start, $token_length );` to produce the entire tag. This patch fixes that, before it would exclude the final `>` on tags.

This is mostly internal and it seems to have adjustments in several places to fix it, but I'd like to have a consistent internal representation. Other types of tokens behave as I'd expect (with `substr( $html, $start, $length )`).

@dmsnell thoughts?

<details>

<summary>I'd expect something like this to hold:</summary>

```php
<?php

$t = '<div class="target">text</div>trailing text';
$start = 0;
$token_length = 20;
$text_start = $token_length - $start;
$text_length = 4;

$end_start = $text_start + $text_length;
$end_length = 6;

var_dump(
    substr( $t, $start, $token_length ),
    substr( $t, $text_start, $text_length ),
    substr( $t, $end_start, $end_length ),
);

```

</details>

Trac ticket: https://core.trac.wordpress.org/ticket/61301

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

File: tests/phpunit/tests/html-api/wpHtmlTagProcessor.php [hunk 1]
```diff
@@ -476,6 +476,49 @@ public function test_get_updated_html_applies_updates_to_content_after_seeking_t
 		$this->assertSame( '<div wonky><img hidden></div>', $processor->get_updated_html() );
 	}
 
+	/**
+	 * Ensures that tags start and length are correctly determined.
+	 *
+	 * @ticket 321
```
Comment (sirreal) [inline tests/phpunit/tests/html-api/wpHtmlTagProcessor.php]: Fix (create) ticket

File: tests/phpunit/tests/html-api/wpHtmlTagProcessor.php [hunk 2]
```diff
@@ -476,6 +476,51 @@ public function test_get_updated_html_applies_updates_to_content_after_seeking_t
...
+
+		$processor->next_tag( array( 'class_name' => 'target' ) );
+
+		$start_property = new ReflectionProperty( $processor, 'token_starts_at' );
+		$start_property->setAccessible( true );
+		$tag_start = $start_property->getValue( $processor );
+
+		$length_property = new ReflectionProperty( $processor, 'token_length' );
+		$length_property->setAccessible( true );
+		$tag_length = $length_property->getValue( $processor );
```
Comment (dmsnell) [inline tests/phpunit/tests/html-api/wpHtmlTagProcessor.php]: I'd rather us avoid reflection in tests, giving tests privileges that normal code shouldn't have.

not sure where, but I had a test set once which verified not the lengths, but the bookmark spans.

```php
$processor = new class ( $html ) extends WP_HTML_Tag_Processor {
	public function get_raw_token() {
		$this->set_bookmark( 'here' );
		$mark = $this->bookmarks['here'];

		return substr( $this->html, $mark->start, $mark->length );
	}
}

for ( $i = 0; $i < $nth_token_in_html; $i++ ) {
	$processor->next_token();
}

$this->assertSame(
	$raw_token,
	$processor->get_raw_token(),
	'Failed to identify full extent of token.'
);
```

**Review by dmsnell — COMMENTED** [review]
This has been a longstanding annoyance to me, as it's clearly wrong, but I've had trouble fixing it without breaking other tokens. Let's see if we can test it from the bookmarks, and then add a comprehensive test dataset with all kinds of tokens and various HTML constructs (e.g. the HTML contains more than just the token by itself so we can see edge cases).

File: tests/phpunit/tests/html-api/wpHtmlTagProcessor.php (hunk 1 above)
Comment (sirreal) [reply tests/phpunit/tests/html-api/wpHtmlTagProcessor.php]:
```suggestion
	 * @ticket 61301
```

File: tests/phpunit/tests/html-api/wpHtmlTagProcessor.php [hunk 3]
```diff
@@ -476,6 +476,78 @@ public function test_get_updated_html_applies_updates_to_content_after_seeking_t
...
+	 *
+	 * @dataProvider data_html_nth_token_substring
+	 *
+	 * @param string $html            Input HTML.
+	 * @param string $match_nth_token The number of times to call ::match_token.
+	 * @param string $expected_match  The expected HTML at the matched token.
+	 */
+	public function test_token_bookmark_span( string $html, int $match_nth_token, string $expected_match ) {
+		$processor = new class( $html ) extends WP_HTML_Tag_Processor {
+			public function get_token_string() {
```
Comment (dmsnell) [inline tests/phpunit/tests/html-api/wpHtmlTagProcessor.php]: not that it matters, but for what it's worth, I've been naming this `get_raw_token()` in multiple other places

File: tests/phpunit/tests/html-api/wpHtmlTagProcessor.php [hunk 4]
```diff
@@ -476,6 +476,78 @@ public function test_get_updated_html_applies_updates_to_content_after_seeking_t
...
+				$this->set_bookmark( 'mark' );
+				$mark = $this->bookmarks['mark'];
+
+				return substr( $this->html, $mark->start, $mark->length );
+			}
+		};
+
+		for ( $i = 0; $i < $match_nth_token; $i++ ) {
+			$processor->next_token(); }
```
Comment (dmsnell) [inline tests/phpunit/tests/html-api/wpHtmlTagProcessor.php]: nitpick: missing newline

File: tests/phpunit/tests/html-api/wpHtmlTagProcessor.php (hunk 3 above)
Comment (dmsnell) [reply tests/phpunit/tests/html-api/wpHtmlTagProcessor.php]: updated

**Comment by dmsnell** [comment]
The test failure in [#6651](https://github.com/WordPress/wordpress-develop/pull/6651) makes me nervous.

Screenshot 2024-05-27 at 4 37 45 PM

**Comment by dmsnell** [comment]
@sirreal can you look into the failure in the linked PR? I think it's related to seeking with updates and getting the wrong `$this->bytes_already_processed`

**Comment by sirreal** [comment]
> The test failure in https://github.com/WordPress/wordpress-develop/pull/6651 makes me nervous.
> Can you look into the failure in the linked PR? I think it's related to seeking with updates and getting the wrong `$this->bytes_already_processed`

68a3251d307c39467ab85bf739578d15da378d75 was missing from that branch. There was a test that used the broken offset. The offset adjustment was removed in this branch, I pushed it to #6651.

**Comment by sirreal** [comment]
@westonruter This is a bugfix for incorrect token lengths in tag bookmarks in the Tag Processor. Subclasses might see and be affected by the bug (and the fix) if they ever work directly with tag bookmarks lengths. I believe you have some plugins that may be doing advanced things with the Tag Processor so wanted to make sure you were aware.

Below is an example fix from this PR, plugins would be impacted in the same way and would need to remove the off-by-one correction on the bookmark lengths:

```diff
diff --git a/src/wp-includes/interactivity-api/class-wp-interactivity-api-directives-processor.php b/src/wp-includes/interactivity-api/class-wp-interactivity-api-directives-processor.php
index 3b2dcb123797..b12dcb4b3b15 100644
--- a/src/wp-includes/interactivity-api/class-wp-interactivity-api-directives-processor.php
+++ b/src/wp-includes/interactivity-api/class-wp-interactivity-api-directives-processor.php
@@ -107,7 +107,7 @@ public function append_content_after_template_tag_closer( string $new_content ):
 
 		$bookmark = 'append_content_after_template_tag_closer';
 		$this->set_bookmark( $bookmark );
-		$after_closing_tag = $this->bookmarks[ $bookmark ]->start + $this->bookmarks[ $bookmark ]->length + 1;
+		$after_closing_tag = $this->bookmarks[ $bookmark ]->start + $this->bookmarks[ $bookmark ]->length;
 		$this->release_bookmark( $bookmark );
 
 		// Appends the new content.
@@ -140,7 +140,7 @@ private function get_after_opener_tag_and_before_closer_tag_positions( bool $rew
 		}
 		list( $opener_tag, $closer_tag ) = $bookmarks;
 
-		$after_opener_tag  = $this->bookmarks[ $opener_tag ]->start + $this->bookmarks[ $opener_tag ]->length + 1;
+		$after_opener_tag  = $this->bookmarks[ $opener_tag ]->start + $this->bookmarks[ $opener_tag ]->length;
 		$before_closer_tag = $this->bookmarks[ $closer_tag ]->start;
 
 		if ( $rewind ) {
```

**Comment by westonruter** [comment]
@sirreal See the following method in [this subclass of `WP_HTML_Tag_Processor`](https://github.com/WordPress/performance/blob/6c27ea6cb2c8bc524bdc2bc4886114456b7f2a88/plugins/optimization-detective/class-od-html-tag-processor.php#L41-L61):

```php
	/**
	 * Appends HTML to the provided bookmark.
	 *
	 * @param string $bookmark Bookmark.
	 * @param string $html     HTML to inject.
	 * @return bool Whether the HTML was appended.
	 */
	public function append_html( string $bookmark, string $html ): bool {
		if ( ! $this->has_bookmark( $bookmark ) ) {
			return false;
		}

		$start = $this->bookmarks[ $bookmark ]->start;

		$this->lexical_updates[] = new WP_HTML_Text_Replacement(
			$start,
			$this->old_text_replacement_signature_needed ? $start : 0,
			$html
		);
		return true;
	}
```

So since it's not dealing with `length` then is a patch needed?

**Comment by dmsnell** [comment]
> So since it's not dealing with length then is a patch needed?

this should be correct, @westonruter

**Review by dmsnell — APPROVED** [review]
Great fix ahead of exposing it in any public interface!

**Comment by dmsnell** [comment]
Merged in [[58233]](https://core.trac.wordpress.org/changeset/58233)
https://github.com/wordpress/wordpress-develop/commit/6ca5bdc3ac193f3aa196b188baab10ec8857c11e

**Pull request closed without merging**